h2o-algos/src/main/java/hex/tree/ExactSplitPoints.java

```diff
@@ -0,0 +1,133 @@
+            }
+            frToTrain[fr.numCols()] = i;
+            fr.add(trainFr.name(i), trainFr.vec(i));
+        }
+        IcedHashSet<IcedDouble>[] values = new ExactSplitPoints(maxCardinality, fr.numCols())
+                .doAll(fr)._values;
+        double[][] splitPoints = new double[trainFr.numCols()][];
+        for (int i = 0; i < values.length; i++) {
+            if (values[i] == null) {
+                continue;
+            }
+            double[] vals = new double[values[i].size()];
+            int valsSize = 0;
+            for (IcedDouble wrapper : values[i]) {
+                vals[valsSize++] = wrapper._val;
+            }
+            assert valsSize == vals.length;
+            Arrays.sort(vals);
+            assert isUniqueSequence(vals);
+            splitPoints[frToTrain[i]] = vals;
+        }
+        return splitPoints;
+    }
+
+    static boolean isUniqueSequence(double[] seq) {
+        if (seq.length == 1)
+            return true;
+        double lastValue = seq[0];
+        for (int i = 1; i < seq.length; i++) {
+            if (lastValue >= seq[i])
+                return false;
+            lastValue = seq[i];
+        }
+        return true;
+    }
+
+    @SuppressWarnings("unchecked")
+    private ExactSplitPoints(int maxCardinality, int nCols) {
+        _maxCardinality = maxCardinality;
+        _values = new IcedHashSet[nCols];
+        for (int i = 0; i < _values.length; i++) {
+            _values[i] = new IcedHashSet<>();
+        }
+    }
+
+    @Override
+    public void map(Chunk[] cs) {
+        Set<IcedDouble> localValues = new HashSet<>(_maxCardinality);
+        for (int col = 0; col < cs.length; col++) {
+            localValues.clear();
+            if (_values[col] == null)
+                continue;
+            Chunk c = cs[col];
+            IcedDouble wrapper = new IcedDouble();
+            for (int i = 0; i < c._len; i++) {
+                double num = c.atd(i);
+                if (Double.isNaN(num))
+                    continue;
+                if (wrapper._val == num)
+                    continue;
+                wrapper.setVal(num);
+                if (localValues.add(wrapper)) {
+                    if (localValues.size() > _maxCardinality) {
+                        _values[col] = null;
+                        break;
+                    }
+                    wrapper = new IcedDouble();
+                }
+            }
+            merge(col, localValues);
+        }
+    }
+
+    private void merge(int col, Collection<IcedDouble> localValues) {
+        final Set<IcedDouble> allValues = _values[col];
+        if (allValues == null)
+            return;
+        allValues.addAll(localValues);
+        if (allValues.size() > _maxCardinality) {
+            _values[col] = null;
+        }
+    }
+
+    @Override
+    public void reduce(ExactSplitPoints mrt) {
+        if (mrt._values != _values) { // merging with a result from a different node
```

Comment:
This condition can be false? I did not expect that the `reduce` method can be called on the same node...

Reply:
reduce can be called on the same node - each 2 maps reduce into each other, however, due to the design of this task - we don't need to execute the same node reduces